homeassistant/components/rainmachine/sensor.py

```diff
@@ -15,33 +16,57 @@
 from homeassistant.core import HomeAssistant, callback
 from homeassistant.helpers.entity import EntityCategory
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util.dt import utcnow
 
 from . import RainMachineEntity
 from .const import (
     DATA_CONTROLLER,
     DATA_COORDINATOR,
     DATA_PROVISION_SETTINGS,
     DATA_RESTRICTIONS_UNIVERSAL,
+    DATA_ZONES,
     DOMAIN,
 )
-from .model import RainMachineSensorDescriptionMixin
+from .model import (
+    RainMachineDescriptionMixinApiCategory,
+    RainMachineDescriptionMixinUid,
+)
+
+DEFAULT_ZONE_COMPLETION_TIME_WOBBLE_TOLERANCE = timedelta(seconds=5)
 
 TYPE_FLOW_SENSOR_CLICK_M3 = "flow_sensor_clicks_cubic_meter"
 TYPE_FLOW_SENSOR_CONSUMED_LITERS = "flow_sensor_consumed_liters"
 TYPE_FLOW_SENSOR_START_INDEX = "flow_sensor_start_index"
 TYPE_FLOW_SENSOR_WATERING_CLICKS = "flow_sensor_watering_clicks"
 TYPE_FREEZE_TEMP = "freeze_protect_temp"
+TYPE_ZONE_RUN_COMPLETION_TIME = "zone_run_completion_time"
+
+ZONE_STATE_NOT_RUNNING = "not_running"
```

Comment:
I know we're using enums more frequently now, but that seemed unnecessary here – open to feedback, though.

Reply:
The enums make sense when there is a high risk of invalid values (ie magically new device classes that HA doesn't know what to do with), unexpected values, or implementation error.  I don't think it is needed here.